src/ws/session.c

```diff
@@ -515,12 +515,13 @@ open_session (pam_handle_t *pamh)
       return res;
     }
 
-  /* Yes, buf "leaks" */
-  buf = malloc (sizeof (struct passwd) + 8192);
-  if (buf == NULL)
+  if (!pwd_buf)
+    pwd_buf = malloc (sizeof (struct passwd) + 8192);
+
+  if (pwd_buf == NULL)
     res = ENOMEM;
   else
-    res = getpwnam_r (name, buf, (char *)(buf + 1), 8192, &pwd);
+    res = getpwnam_r (name, pwd_buf, (char *)(pwd_buf + 1), 8192, &pwd);
```

Comment:
OK, if you don't want to change it from a pointer to a struct, then maybe just havea a `static struct pwd _pwd;` in `open_session()` that you use as the first argument of `getpwnam_r()? `pwd` will just point to `_pwd` then, so that makes the patch less intrusive but still gives type safety.